server/src/main/kotlin/suwayomi/tachidesk/manga/impl/update/Updater.kt

```diff
@@ -0,0 +1,72 @@
+package suwayomi.tachidesk.manga.impl.update
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import mu.KotlinLogging
+import suwayomi.tachidesk.manga.impl.Chapter
+import suwayomi.tachidesk.manga.model.dataclass.MangaDataClass
+
+class Updater : IUpdater {
+    private val logger = KotlinLogging.logger {}
+    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
+
+    private var isRunning: Boolean = false
+
+    private var tracker = HashMap<String, UpdateJob>()
+    private var updateChannel = Channel<UpdateJob>()
+    private val statusChannel = MutableStateFlow(UpdateStatus())
+    private var updateJob: Job? = null
+
+    init {
+        updateJob = createUpdateJob()
+    }
+
+    private fun createUpdateJob(): Job {
+        return scope.launch {
+            while (true) {
+                val job = updateChannel.receive()
+                process(job)
+                isRunning = !updateChannel.isEmpty
+                val value = UpdateStatus(tracker, isRunning)
+                statusChannel.emit(value)
+            }
+        }
+    }
+
+    private suspend fun process(job: UpdateJob) {
+        isRunning = true
+        job.status = JobStatus.RUNNING
+        tracker["${job.manga.id}"] = job
+        statusChannel.emit(UpdateStatus(tracker, isRunning))
+        try {
+            logger.info { "Updating ${job.manga.title}" }
+            Chapter.getChapterList(job.manga.id, true)
+            job.status = JobStatus.COMPLETE
+        } catch (e: Exception) {
```

Comment:
Re-throw cancellation exceptions here

Reply:
Wouldn't that cancel the whole job? My intention here is to accept the error and mark the specific manga as failed.

Reply:
`CancellationException`'s are thrown when the job(thread) is cancelled, so they need to be rethrown in any place where you catch exceptions

Reply:
Like so
```kotlin
try {
    println(channel.receive())
} catch(e: Exception) {
    if (e is CancellationException) throw e
    // log error
}
```